runtime/src/main/java/org/corfudb/runtime/collections/streaming/StreamPollingScheduler.java

```diff
@@ -0,0 +1,321 @@
+            } catch (InterruptedException ie) {
+                log.warn("Scheduler thread has been interrupted!", ie);
+                Thread.currentThread().interrupt();
+            } catch (Throwable t) {
+                log.error("Unexpected throwable!", t);
+            }
+        }
+    }
+
+    private void tick() {
+        scheduler.submit(new Tick());
+    }
+
+    public void addTask(@Nonnull StreamListener streamListener, @Nonnull String namespace,
+                        @Nonnull String streamTag, @Nonnull List<String> tablesOfInterest,
+                        long lastAddress, int bufferSize) {
+
+        Preconditions.checkArgument(bufferSize >= pollThreshold);
+        synchronized (allTasks) {
+            if (allTasks.containsKey(streamListener)) {
+                // Multiple subscribers subscribing to same namespace and table is allowed
+                // as long as the hashcode() and equals() method of the listeners are different.
+                throw new StreamingException(
+                        "StreamingManager::subscribe: listener already registered " + streamListener);
+            }
+            StreamingTask task = new StreamingTask(runtime, workers, namespace, streamTag, streamListener,
+                    tablesOfInterest, lastAddress, bufferSize);
+            allTasks.put(streamListener, task);
+            log.debug("addTask: added {} {} {}", streamListener, namespace, streamTag);
+            allTasks.notifyAll();
+        }
+    }
+
+    public void removeTask(@Nonnull StreamListener streamListener) {
+        synchronized (allTasks) {
+            allTasks.remove(streamListener);
+            allTasks.notifyAll();
+        }
+    }
+
+    // hole filling mitigation
+    // sequencer regression mitigation
+    // fairness?
+
+    private void waitForTasks() throws InterruptedException {
+        synchronized (allTasks) {
+            while (allTasks.isEmpty()) {
+                allTasks.wait();
+            }
+        }
+    }
+
+    private List<StreamingTask> getTasks(StreamStatus status, Predicate<StreamingTask> predicate) {
+        List<StreamingTask> result = new ArrayList<>();
+        for (StreamingTask task : allTasks.values()) {
+            if (task.getStatus() == status && predicate.test(task)) {
+                result.add(task);
+            }
+        }
+        return result;
+    }
+
+    private List<StreamingTask> getTasks(StreamStatus status) {
+        return getTasks(status, t -> true);
+    }
+
+    private List<StreamAddressRange> getPollQueries(List<StreamingTask> tasks) {
+        List<StreamAddressRange> pollRequests = new ArrayList<>(tasks.size());
+        for (StreamingTask task : tasks) {
+            DeltaStream stream = task.getStream();
+            pollRequests.add(new StreamAddressRange(stream.getStreamId(), Address.MAX, stream.getMaxAddressSeen()));
+        }
+        return pollRequests;
+    }
+
+    private void poll(List<StreamingTask> tasks) {
```

Comment:
can we add comments? took me a while to internalize that 'poll' meant sequencer poll and not data polling (actual read of data, which I believe is what you have called sync)

Reply:
Done.